src/coaster.cpp

```diff
@@ -721,17 +721,19 @@ int CoasterInstance::GetFirstPlacedTrackPiece() const
 	return -1;
 }
 
-std::vector<XYZPoint16> CoasterInstance::get_all_positions() {
+std::vector<XYZPoint16> CoasterInstance::GetAllPiecePositions() {
 	std::vector<XYZPoint16> result;
 
 	const int start_piece = GetFirstPlacedTrackPiece();
 	if (start_piece < 0) return result;
 
-	result.push_back(this->pieces[start_piece].base_voxel);
+	result.emplace_back(this->pieces[start_piece].base_voxel);
 	for (int p = start_piece;;) {
 		p = FindSuccessorPiece(this->pieces[p]);
 		if (p < 0 || p == start_piece) break;
-		result.push_back(this->pieces[p].base_voxel);
+		for (const TrackVoxel *subpiece : this->pieces[p].piece->track_voxels) {
+			result.emplace_back(this->pieces[p].base_voxel + subpiece->dxyz);
+		}
```

Comment:
Sorry, but after a night of sleep this bothered me. It looks like the wrong approach to me, since you're breaking encapsulation.

A `Coaster` has `Trackpiece`s, and a `CoasterInstance` has `PositionedTrackPiece`s. Those are the classes that know internals of track pieces. Here you export that knowledge to `CoasterInstance` level.

It may look like a simple solution, but now, track piece internals are not just in track pieces, they are also here. So if you change track piece internals say one year after now, you have to change two places now. There is a large chance you are not going to keep those two spots consistent.